### ISSUE TITLE
Marshal: Original is inefficient

### ISSUE THREAD

**Issue by arv** [attic-labs/noms#2990]
Sorry for not catching this in the review.

When using `noms:",original"` in Marshal we start with the original struct and then repeatedly calls `Set`. Each `Set` copies creates a new struct and copies all the fields over. It would be better to assembly all the fields once and then create the type.

**Comment by arv** [comment]
This also conflicts with #2972. @kalman

**Comment by aboodman** [comment]
Yeah I actually called this out in the code. It is surprisingly complicated because the struct can have a user-provided type that is different from what type accretion would calculate.

So the way it needs to work is that you start with the StructData from the original, and then add into that all the fields from the Go type. So you need to decide what happens here:

```
base := types.NewStructFromFields("S", types.StructData{
  "foo": types.Value,
});

type S struct {
  Foo: int
  Original: types.Struct `noms:",original"`
}

unmarshaled := S{42, base}
marshaled := types.Marshal(unmarshaled)
```

`marshaled.foo` needs to end up being `types.Value`.
